pinctrl: esp32/s2/c3: initial support

glaubermaroto opened the pull request:
This PR brings `pinctrl` support to Espressif SoCs and their related boards.

It also removes `pinmux` support off Espressif SoCs.

This work links to #39740.

Closes #43715

Commit:
drivers: pinctrl: esp32: initial support

add initial pinctrl driver support for ESP32.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp32: dts: uart: add default UART pinctrl states

through the inclusion of pinctrl-[0/name] properties to
ESP32's board UART nodes.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
dts: esp32: add pinctrl bindings and definitions

to support implementation and peripheral usage of the pinctrl
driver API.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
drivers: serial: esp32: use pinctrl API

and removes references/usage of old pinmux properties.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
dts: bindings: serial: esp32: add pinctrl properties

and removes bindings used by pinmux.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp32: dts: spi: add default SPI pinctrl states

through the inclusion of pinctrl-[0/name] properties to
ESP32's board SPI nodes.

It also removes dts properties related to pinmux.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
dts: bindings: spi: esp32: add pinctrl properties

and removes bindings previously used by the
pinmux driver.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
drivers: spi: esp32: use pinctrl API

and removes references/usage of old pinmux properties.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: dts: spi: remove property "use-iomux"

which became deprecated when pinctrl came into stage.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp32: dts: i2c: add default I2C pinctrl states

Through the inclusion of pinctrl-[0/name] properties to
ESP32's board I2C nodes.

Also removes the i2c1 node, which currently is not being
used by any sample code or test case.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: dts: bindings: i2c: add pinctrl properties

and removes bindings previously used by the
pinmux driver.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: dts: bindings: i2c: update pin properties

After pinctrl's subsystem support, I2C pin properties are
no longer required for pin muxing, however, this information
is still valuable in an eventual FSM failure if the target
SoC has no harwared mechanisms to support bus recovery.

In this case, bus recovery uses pin information to restore
the I2C bus state.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
dts: xtensa: esp32s2: add pinctrl node and bindings

to support implementation and peripheral usage of the pinctrl
driver API.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
drivers: pinctrl: esp32s2: initial support

add initial pinctrl driver support for ESP32S2.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp32s2_saola: dts: default pinctrl states

Add default pinctrl states to UART, SPI and I2C
peripherals.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32s2: dts: remove property "use-iomux"

which became deprecated when pinctrl came into stage.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32s2: drivers: spi: removal of unused definitions

Clean up of pin signals definitions previously used
by the pinmux driver, now deprecated by the use of
the pinctrl API.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32c3: drivers: pinctrl: initial support

add initial pinctrl driver support for ESP32C3.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32c3: dts: add pinctrl node and bindings

to support implementation and peripheral usage of the pinctrl
driver API.

Removes deprecated "use-iomux" property from the SPI node.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp32c3_devkitm: dts: default pinctrl states

Add default pinctrl states to UART, SPI and I2C
peripherals.

I2C's pin information information is kept to
trigger the bus recovery mechanism in case of
failure.

I2C_0's default SDA pin is changed to no conflict
with SPI2's MISO pin.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32c3: drivers: spi: removal of unused definitions

Clean up of pin signals definitions previously used
by the pinmux driver, now deprecated by the use of
the pinctrl API.

Refactor device tree macros usage to make usage of
SPI instances more general.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp_wrover_kit: dts: add default pinctrl states

update peripheral nodes to use default pinctrl states.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: drivers: gpio: remove pinmux dependency

Refactor the GPIO driver code to remove dependency
from the pinmux API.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: heltec_wifi_lora32_v2: dts: add default pinctrl states

update peripheral nodes to use default pinctrl states.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: driver: pinmux: driver removal

Removal of ESP32's pinmux driver since it
was deprecated by pinctrl.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: dts: bindings: removal of pinmux bindings

those bindings are no more needed since pinmux was
deprecated in favor of pinctrl.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32/s2/c3-based boards: remove pinmux definitions

ESP32/S2/C3-based boards no longer support pinmux, which was
deprecated in favor of pinctrl.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
CODEOWNERS: pinmux & pinctrl: update ownership

- remove myself as *esp32* pinmux code owner
- add myself as *esp32* pinctrl code owner

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp_wrover_kit: update CMakeLists.txt

remove dependency on pinmux definition.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: heltec_wifi_lora32_v2: update CMakeLists.txt

remove dependency on pinmux definition.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: odroid_go: dts: add default pinctrl states

update peripheral nodes to use default pinctrl states.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: esp_wrover_kit: change init priority

making board initialization depend on
GPIO_INIT_PRIORITY and setting the
initialization level to APPLICATION.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: heltec_wifi_lora32_v2: change init priority

making board initialization depend on
GPIO_INIT_PRIORITY.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32/s2/c3: dts: remove pinmux node

remove the pinmux node from ESP32, ESP32-S2
and ESP32-C3 SoCs.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

gmarull commented on drivers/pinctrl/pinctrl_esp32.c:
```diff
@@ -0,0 +1,123 @@
+
+	port_device = gpio_ports[ESP32_PORT_IDX(pin_num)];
+
+	if ((port_device == NULL) || (!device_is_ready(port_device))) {
+		return -ENODEV;
+	}
```
please don't mix GPIO/pinctrl. pinctrl drivers need to be independent, and, if there's ever a relation it'll likely be GPIO depending on pinctrl (since in most cases it's a subset). Same discussion happened for the Microchip case, see https://github.com/zephyrproject-rtos/zephyr/pull/43282

glaubermaroto replied:
@gmarull thanks for pointing that out. I have decoupled the pinctrl driver from the GPIO driver, removing dependency between them.

Though you will see some gpio-prefixed naming are kept here and there, it is due to hal_espressif's own conventions.

gmarull replied:
I still see the gpio based implementation on the first commit, can you please squash//fixup?

glaubermaroto replied:
Squashed/Fixed.

gmarull commented on boards/xtensa/esp32/esp32_defconfig:
```diff
@@ -18,6 +18,8 @@ CONFIG_XTENSA_USE_CORE_CRT1=n
 CONFIG_PINMUX=y
 CONFIG_PINMUX_ESP32=y
 
+CONFIG_PINCTRL=y
```
suggestion: default PINCTRL to y at SoC level if it's always required.

glaubermaroto replied:
Moved to SoC level. Thanks for the suggestion.

gmarull commented on drivers/pinctrl/Kconfig.esp32:
```diff
@@ -0,0 +1,9 @@
+config PINCTRL_ESP32
+	bool "Pin controller driver for ESP32 SoC"
+	depends on SOC_ESP32
+	default y
```
please default based on `dt_compat_status`

glaubermaroto replied:
Done.

gmarull replied:
can't see this fixed in the first commit, can you fixup this one?

glaubermaroto replied:
Fixed.

gmarull commented on dts/bindings/pinctrl/espressif,esp32-pinctrl.yaml:
```diff
@@ -0,0 +1,56 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    ESP32 Pin controller Node
+    Based on pincfg-node.yaml binding.
+
+    Note: `bias-disable` and `drive-push-pull` are default pin configurations.
+           They will be applied in case no `bias-foo` or `driver-bar` properties
+           are set.
```
please extend description on usage (see other vendors as a reference, e.g. GD32, nRF, Microchip, etc.)

gmarull commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -676,6 +678,9 @@ static int IRAM_ATTR i2c_esp32_init(const struct device *dev)
 
 #define ESP32_I2C_INIT(idx)		\
+					\
+PINCTRL_DT_INST_DEFINE(idx);		\
```
nit: please align with other macro content

glaubermaroto replied:
This driver will be addressed again in another PR, the scope of that PR also includes some code refactoring.

If you don't mind, I'd rather address this alignment suggestion in that PR to not cloud pinctrl-related modifications in here, ok?

gmarull commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -41,6 +43,14 @@ LOG_MODULE_REGISTER(i2c_esp32, CONFIG_I2C_LOG_LEVEL);
 
+#define I2C_GET_PIN_INFO(state, idx)								\
+{												\
+	.gpio_name = gpio_esp32_get_gpio_for_pin(ESP32_PIN_NUM(state->pins[idx].pinmux)),	\
+	.sig_out = ESP32_PIN_SIGO(state->pins[idx].pinmux),					\
+	.sig_in = ESP32_PIN_SIGI(state->pins[idx].pinmux),					\
+	.pin = ESP32_PIN_NUM(state->pins[idx].pinmux)						\
+}
```
it is not allowed to query pin information, API operates on states. all pinctrl structs should be considered private.

glaubermaroto replied:
This query is not part of the regular flow of the pinctrl API. It is included here to overcome an I2C driver constraint:

Most Espressif SoCs are provisioned with HAL mechanisms to overcome an eventual I2C FSM failure, while others need to keep pin numbering info to achieve the same.

gmarull replied:
The problem here is that driver's can't make use of any pinctrl state information since it is opaque. You can't rely on the data structure behind pinctrl_dev_config. If you need GPIO access to i2c pins, you'll need something like:

1. Create a new custom state "gpio" (or whatever name) that configures pins as GPIO instead of muxed to the I2C peripheral
2. Store GPIO phandles in DT, e.g. `i2c-gpios = <&gpiox 0 1>, <&gpioy 2 3>;`. It's true that this duplicates some information also stored by pinctrl, but right now, the pinctrl API doesn't expose any specified pin as a GPIO device. We may explore this in the future based on the initial round of pinctrl ports.
3. When in need for gpio access, just do `pinctrl_apply_state(..., PINCTRL_GPIO_STATE);`. Use GPIO API. When done, apply `PINCTRL_DEFAULT_STATE` back.

glaubermaroto replied:
I have gone through a different approach but got rid of the pin state query.

gmarull reviewed (changes requested):
I've done a quick initial review. Please check my comments.

glaubermaroto commented:
@carlescufi Would you please add me as one of the assignees at #39740?

Could you also help me with this [failing compliance check](https://github.com/zephyrproject-rtos/zephyr/runs/5479786024?check_suite_focus=true)? My understanding is that using `typedef` to define `pinctrl_soc_pin_t` should be an exception to that rule, right?

glaubermaroto commented:
@adalava Could you gently check/report any misbehavior when testing this PR on Heltec's LoRa board? Feel free to get in touch.

gmarull commented:
> Could you also help me with this [failing compliance check](https://github.com/zephyrproject-rtos/zephyr/runs/5479786024?check_suite_focus=true)? My understanding is that using `typedef` to define `pinctrl_soc_pin_t` should be an exception to that rule, right?

This failure is expected, you can ignore it.

Commit:
soc: esp32/s2/c3: make PINCTRL config default

through the selection of PINCTRL config
at the SoC level.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

adalava commented:
> @adalava Could you gently check/report any misbehavior when testing this PR on Heltec's LoRa board? Feel free to get in touch.

@glaubermaroto , I tested basic/button and subsys/display/lvgl samples with your branch, they exercise a lot of pin stuff. It looks good to me!

sylvioalves commented on drivers/serial/uart_esp32.c:
```diff
@@ -530,35 +466,14 @@ static const DRAM_ATTR struct uart_driver_api uart_esp32_api = {
+PINCTRL_DT_INST_DEFINE(idx);							\
+										\
 static const DRAM_ATTR struct uart_esp32_config uart_esp32_cfg_port_##idx = {	       \
 	.uart_num = DT_INST_PROP(idx, peripheral),	\
```
Remove `uart_num`, it is not necessary anymore.
Also, remove `peripheral` from devicetree binding so #43715 can be closed.

glaubermaroto replied:
@sylvioalves Removal was done here, I'll open another PR to address `hal_espressif`'s side.

Commit:
esp32: drivers: uart: remove uart_num

When using pin states, uart_num is no
longer required.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32/s2/c3: dts: uart: remove peripheral property

The pinctrl support made usage of the 'peripheral'
property no longer required.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented:
Twister is [raising an error](https://github.com/zephyrproject-rtos/zephyr/runs/5527942652?check_suite_focus=true) when running on the odroig_go board, however, this board is not being maintained by Espressif.

I'm not sure how should be the approach to this. Not including the odroid_go board on this PR, in turn, will lead to compilation errors.

@nashif @sylvioalves Could you help me with this one?

glaubermaroto commented:
The last push was to solve file conflicts.

gmarull commented on drivers/pinctrl/pinctrl_esp32.c:
```diff
@@ -0,0 +1,270 @@
+		return -EINVAL;
+	}
+
+	port_device = gpio_ports[ESP32_PORT_IDX(pin_num)];
+
+	if ((port_device == NULL) || (!device_is_ready(port_device))) {
+		return -ENODEV;
+	}
```
this driver continues depending on GPIO, please make driver independent of GPIO. See e.g. GD32 driver where GPIO information is used from DT, but no GPIO device is used internally.

glaubermaroto replied:
Fixed.

glaubermaroto commented:
> Twister is [raising an error](https://github.com/zephyrproject-rtos/zephyr/runs/5527942652?check_suite_focus=true) when running on the odroid_go board, however, this board is not being maintained by Espressif.
> 
> I'm not sure how should be the approach to this. Not including the odroid_go board on this PR, in turn, will lead to compilation errors.
> 
> @nashif @sylvioalves Could you help me with this one?

gmarull commented on boards/xtensa/heltec_wifi_lora32_v2/CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-if(CONFIG_PINMUX_ESP32)
+if(CONFIG_PINCTRL_ESP32)
```
seems like board is using just GPIO API to configure some pins, how is that related to pinctrl?

glaubermaroto replied:
Good point. I have changed it to evaluate `CONFIG_GPIO_ESP32` instead.

gmarull commented on boards/xtensa/esp_wrover_kit/CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-if(CONFIG_PINMUX_ESP32)
+if(CONFIG_PINCTRL_ESP32)
```
ditto

glaubermaroto replied:
Good point. I have changed it to evaluate `CONFIG_GPIO_ESP32` instead.

gmarull reviewed (changes requested):
Thanks for the continued work, I have some questions please check.

sylvioalves commented:
rebased to check odroid_go twister issue.

Commit:
esp32: drivers: i2c: use pinctrl

though using pinctrl's subsystem, the I2C driver
keeps pin information in case of communication
failure. This information is needed in case of
FSM failure.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32s2: drivers: i2c: tweak FSM-related logic

Differently from ESP32, the ESP32-S2 SoC has native
hal support to reset its own I2C FSM in case of failure.
This commit removes warnings related to unused reset
logic, which does not really apply to ESP32-S2.

It also removes code and data structures related to pin
information from the build when the target SoC supports
hardware mechanisms to reset the I2C FSM.

Finally, it checks at compile time if the preconditions
for correct bus recovery are being met.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: odroid_go: change init priority

making board initialization depend on
GPIO_INIT_PRIORITY and setting the
initialization level to APPLICATION.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: odroid_go: update CMakeLists.txt

remove dependency on pinmux definition.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented:
> > @adalava Could you gently check/report any misbehavior when testing this PR on Heltec's LoRa board? Feel free to get in touch.
> 
> @glaubermaroto , I tested basic/button and subsys/display/lvgl samples with your branch, they exercise a lot of pin stuff. It looks good to me!

@adalava It might be worth re-testing on Heltec's board again, I have made a couple of changes here and there.

glaubermaroto commented:
@ydamigos Could you try this PR on the `odroid_go` board? Please, let me know how it goes on that board.

ydamigos commented:
@glaubermaroto I did a quick check and got compilation error about missing "espressif/32/esp32-pinctrl.dtsi" file. I' ll have a better look.

glaubermaroto commented:
> @glaubermaroto I did a quick check and got compilation error about missing "espressif/32/esp32-pinctrl.dtsi" file. I' ll have a better look.

That's on `hal_espressif`'s side. Before building, you probably need to run these two:

```sh
west update
west espressif update
```

ydamigos commented:
> That file is on `hal_espressif`'s side. Before building, you probably need to run these two commands:
> 
> ```shell
> west update
> west espressif update
> ```

That did the trick. UART and GPIO output work fine. Tested with hello_world sample and blinky.
GPIO input didn't work using button sample. I will check if this issue is related to this PR or not.

adalava commented:
> > @adalava It might be worth re-testing on Heltec's board again, I have made a couple of changes here and there.
> 
> @glaubermaroto, the button sample is fine but the display sample is not working anymore after latest changes, the display doesn't turn on. I can make it work manually by simply wiring pin16 (OLED RESET) to 3V3 directly.
> 
> To help you reproduce it, here's where I've set the pin16 to HIGH during board initialization: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/xtensa/heltec_wifi_lora32_v2/board_init.c#L28. I think you can do something similar on your board and check voltage.
> 
> Build command: `west build -p auto -b heltec_wifi_lora32_v2 samples/subsys/display/lvgl -- -DSHIELD=ssd1306_128x64`

Hi @glaubermaroto, looks like it's just the order of the things. In boards/xtensa/heltec_wifi_lora32_v2/board_init.c  I switched SYS_INIT back to PRE_KERNEL_2 and display sample is back working fine to me :)

glaubermaroto commented:
> > @glaubermaroto, the button sample is fine but the display sample is not working anymore after latest changes, the display doesn't turn on. I can make it work manually by simply wiring pin16 (OLED RESET) to 3V3 directly.
> 
> Hi @glaubermaroto, looks like it's just the order of the things. In boards/xtensa/heltec_wifi_lora32_v2/board_init.c I switched SYS_INIT back to PRE_KERNEL_2 and display sample is back working fine to me :)

Right, thanks for pointing this out. I'll revert the board's initialization priority before moving on.

glaubermaroto commented:
For now, I have changed this PR to Draft/DNM in order to address some review points.

ydamigos commented:
> @ydamigos Something similar might be happening to the odroid_go board. Could you confirm?

@glaubermaroto The issue was related to missing pull up. I thought all the buttons had external pull up but that's not the case. I' ll open a PR to address it.

Commit:
west.yml: update hal_espressif's revision

update hal_espressif's revision to latest
HAL pinctrl support

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented on dts/bindings/pinctrl/espressif,esp32-pinctrl.yaml:
```diff
@@ -0,0 +1,116 @@
+          drive-open-drain;
+          input-enable;
+          output-high;
+      };
```
I have changed the approach here, only pin muxing information is pre-defined now. Pin properties are selected on-demand.

gmarull commented on dts/bindings/pinctrl/espressif,esp32-pinctrl.yaml:
```diff
@@ -0,0 +1,161 @@
+
+        i2c0_sda_gpio21: i2c0_sda_gpio21 {
+                pinmux = <I2C0_SDA_GPIO21>;
+                bias-pull-up;
+                drive-open-drain;
+                output-high;
```
nit: is this still correct (ie i2c pull-up)?

glaubermaroto replied:
The difference now is that - differently from the previous version of this PR - using this and other properties is no longer mandatory, but the option was kept in case one's wants to use SoC's internal pull-up at its own discretion.

gmarull replied:
in general internall pull-ups are not suitable for I2C (even though in many cases "it just works" for certain configurations).

gmarull commented on include/dt-bindings/pinctrl/esp-pinctrl-common.h:
```diff
@@ -0,0 +1,72 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_ESP_PINCTRL_COMMON_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_ESP_PINCTRL_COMMON_H_
+
+#define ESP32_PIN_NUM_SHIFT      0U
```
nit: this header makes use of `BIT()`, so it should include `<sys/util_macro.h>`

glaubermaroto replied:
In this case, "BIT()" is one of those macros that is also defined on Espressif's HAL.

gmarull replied:
I mean, header is not self-contained, this is a bad practice.

gmarull reviewed (approved):
LGTM, just have a couple of minor comments

gmarull commented:
@glaubermaroto not sure if it's the case, but please also check in samples/tests if there's any overlay changing pins for ESP32 boards.

glaubermaroto commented:
> @glaubermaroto not sure if it's the case, but please also check in samples/tests if there's any overlay changing pins for ESP32 boards.

I've checked it previously. Anyway, thanks for drawing attention to this.

ydamigos commented on dts/xtensa/espressif/esp32.dtsi:
```diff
@@ -9,6 +9,7 @@
 #include <dt-bindings/i2c/i2c.h>
 #include <dt-bindings/clock/esp32_clock.h>
 #include <dt-bindings/interrupt-controller/esp-xtensa-intmux.h>
+#include <dt-bindings/pinctrl/esp32-pinctrl.h>
```
I tried to build lvgl sample for odroid_go but this file is missing.

glaubermaroto replied:
@ydamigos I am facing no issues when building this sample for the odroid_go board. Did you try the following commands before?

```
west update
west espressif update
```

ydamigos replied:
False alarm. I always forget to run these commands first.

glaubermaroto replied:
Something similar was reported on Heltec's board a while ago. Could you try editing `boards/xtensa/odroid_go/board_init.c` and replace `SYS_INIT`'s second argument by `PRE_KERNEL_2` instead of `APPLICATION`?

ydamigos replied:
I did the change but it didn't work. I tried also `samples/drivers/display` with the same result.

ydamigos replied:
I agree., it's really strange. I checked odroid_go device tree files several times to find out what is wrong without success. I will try to create a Discord account the following days and ping you there.

gmarull reviewed (changes requested):
These overlays need to be updated:

samples/bluetooth/hci_uart/boards/esp32.overlay
samples/subsys/fs/fat_fs/boards/esp_wrover_kit.overlay

glaubermaroto commented:
> These overlays need to be updated:
> 
> samples/bluetooth/hci_uart/boards/esp32.overlay samples/subsys/fs/fat_fs/boards/esp_wrover_kit.overlay

That's true. I have grepped only inside `samples/drivers`

Commit:
esp32: samples: overlays: use pin states

Updates samples' overlays to use pin states.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

adalava commented:
@glaubermaroto I verified samples/subsys/display/lvgl and samples/basic/button/ at 75f568a452f6b3ed91f8fb39fcccb1e3d0371fd4 on board heltec_wifi_lora32_v2, it's everything working as expected!

ydamigos commented on boards/xtensa/odroid_go/odroid_go-pinctrl.dtsi:
```diff
@@ -0,0 +1,53 @@
+	};
+
+	i2c0_scl_gpio15: i2c0_scl_gpio15 {
+		pinmux = <I2C0_SCL_GPIO22>;
```
I2C0_SCL_GPIO15

glaubermaroto replied:
Fixed, thanks.

ydamigos commented on drivers/spi/spi_esp32_spim.c:
```diff
@@ -419,7 +332,12 @@ static const struct spi_driver_api spi_api = {
 #endif
 
+#define SPI2_PINSTATE_IDX	0
+#define SPI3_PINSTATE_IDX	1
```
these defines doesn't work properly when board has one SPI instance (like odroid_go only has SPI3).

mbolivar-nordic replied:
Perhaps related to what @ydamigos is saying here, this does appear to be using instance numbers incorrectly.

Please read https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_INST carefully, especially these parts:

```
Instance numbers have the following properties:

- for each compatible, instance numbers start at 0 and are contiguous
- exactly one instance number is assigned for each node with a compatible,
  **including disabled nodes**
- enabled nodes (status property is "okay" or missing) are assigned the
  instance numbers starting from 0, and disabled nodes have instance
  numbers which are greater than those of any enabled node

No other guarantees are made. In particular:

- instance numbers **in no way reflect** any numbering scheme that
  might exist in SoC documentation, node labels or unit addresses,
  or properties of the /aliases node (use DT_NODELABEL() or DT_ALIAS()
  for those)
- there **is no general guarantee** that the same node will have
  the same instance number between builds, even if you are building
  the same application again in the same build directory
```

You seem to be making some assumptions here contrary to what the API guarantees. Please adjust the PR accordingly.

ydamigos replied:
@mbolivar-nordic yes, the issue I am referring to is related with the wrong SPI instance numbers.   
I tried the following changes based on other SPI driver implementations and it works fine:

```
diff --git a/drivers/spi/spi_esp32_spim.c b/drivers/spi/spi_esp32_spim.c
index ae6d249b27..f018c2331f 100644
--- a/drivers/spi/spi_esp32_spim.c
+++ b/drivers/spi/spi_esp32_spim.c
@@ -327,57 +327,48 @@ static const struct spi_driver_api spi_api = {
 };
 
 #ifdef CONFIG_SOC_ESP32
-#define GET_AS_CS(idx) .as_cs = DT_PROP(DT_NODELABEL(spi##idx), clk_as_cs),
+#define GET_AS_CS(idx) .as_cs = DT_INST_PROP(idx, clk_as_cs),
 #else
 #define GET_AS_CS(idx)
 #endif
 
-#define SPI2_PINSTATE_IDX	0
-#define SPI3_PINSTATE_IDX	1
-
 #define ESP32_SPI_INIT(idx)	\
 				\
-PINCTRL_DT_INST_DEFINE(SPI##idx##_PINSTATE_IDX);	\
+PINCTRL_DT_INST_DEFINE(idx);	\
 										\
 	static struct spi_esp32_data spi_data_##idx = {	\
 		SPI_CONTEXT_INIT_LOCK(spi_data_##idx, ctx),	\
 		SPI_CONTEXT_INIT_SYNC(spi_data_##idx, ctx),	\
-		SPI_CONTEXT_CS_GPIOS_INITIALIZE(DT_NODELABEL(spi##idx), ctx)	\
+		SPI_CONTEXT_CS_GPIOS_INITIALIZE(DT_DRV_INST(idx), ctx)	\
 		.hal = {	\
-			.hw = (spi_dev_t *)DT_REG_ADDR(DT_NODELABEL(spi##idx)),	\
+			.hw = (spi_dev_t *)DT_INST_REG_ADDR(idx),	\
 		},	\
 		.dev_config = {	\
-			.half_duplex = DT_PROP(DT_NODELABEL(spi##idx), half_duplex),	\
+			.half_duplex = DT_INST_PROP(idx, half_duplex),	\
 			GET_AS_CS(idx)							\
-			.positive_cs = DT_PROP(DT_NODELABEL(spi##idx), positive_cs),	\
-			.no_compensate = DT_PROP(DT_NODELABEL(spi##idx), dummy_comp),	\
-			.sio = DT_PROP(DT_NODELABEL(spi##idx), sio)	\
+			.positive_cs = DT_INST_PROP(idx, positive_cs),	\
+			.no_compensate = DT_INST_PROP(idx, dummy_comp),	\
+			.sio = DT_INST_PROP(idx, sio)	\
 		}	\
 	};	\
 		\
 	static const struct spi_esp32_config spi_config_##idx = {	\
-		.spi = (spi_dev_t *)DT_REG_ADDR(DT_NODELABEL(spi##idx)),	\
+		.spi = (spi_dev_t *)DT_INST_REG_ADDR(idx),	\
 			\
-		.clock_dev = DEVICE_DT_GET(DT_CLOCKS_CTLR(DT_NODELABEL(spi##idx))),	\
+		.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(idx)),	\
 		.duty_cycle = 0, \
 		.input_delay_ns = 0, \
-		.irq_source = DT_IRQN(DT_NODELABEL(spi##idx)), \
-		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(SPI##idx##_PINSTATE_IDX),	\
+		.irq_source = DT_INST_IRQN(idx), \
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
 		.clock_subsys =	\
-			(clock_control_subsys_t)DT_CLOCKS_CELL(	\
-				DT_NODELABEL(spi##idx), offset),	\
+			(clock_control_subsys_t)DT_INST_CLOCKS_CELL(	\
+				idx, offset),	\
 					\
 	};	\
 		\
-	DEVICE_DT_DEFINE(DT_NODELABEL(spi##idx), &spi_esp32_init,	\
+	DEVICE_DT_INST_DEFINE(idx, &spi_esp32_init,	\
 			      NULL, &spi_data_##idx,	\
 			      &spi_config_##idx, POST_KERNEL,	\
 			      CONFIG_SPI_INIT_PRIORITY, &spi_api);
 
-#if DT_NODE_HAS_STATUS(DT_NODELABEL(spi2), okay)
-ESP32_SPI_INIT(2);
-#endif
-
-#if DT_NODE_HAS_STATUS(DT_NODELABEL(spi3), okay)
-ESP32_SPI_INIT(3);
-#endif
+DT_INST_FOREACH_STATUS_OKAY(ESP32_SPI_INIT)
```

gmarull commented:
compliance failure can be ignored, it's for the `pinctrl_soc_pin_t` typedef.

mbolivar-nordic reviewed (dismissed):
Please revisit the use of devicetree instance numbers; this does not seem correct to me.

I am leaving a comment on one patch, but the same check should be performed on all the driver changes in this PR.

glaubermaroto commented:
Rebased to solve conflicts.

glaubermaroto commented:
> Please revisit the use of devicetree instance numbers; this does not seem correct to me.
> 
> I am leaving a comment on one patch, but the same check should be performed on all the driver changes in this PR.

@mbolivar-nordic I have updated the usage of DT instance numbers on the SPI driver, please, take another look.

glaubermaroto commented:
I am not sure why but one of the approvals on this PR is not being validated.

glaubermaroto commented:
It was mentioned before by @gmarull on this PR but just as a refresher for whoever is able to merge it:

The compliance failure can be ignored, it's for the `pinctrl_soc_pin_t` typedef.

glaubermaroto commented:
@carlescufi This PR is ready. How should we proceed? Can we move forward with it?

mbolivar-nordic commented:
> > I am leaving a comment on one patch, but the same check should be performed on all the driver changes in this PR.
> 
> @mbolivar-nordic I have updated the usage of DT instance numbers on the SPI driver, please, take another look.

@glaubermaroto you seem to have missed the comment that "the same check should be performed on all the driver changes in this PR"

gmarull commented on drivers/pinctrl/pinctrl_esp32.c:
```diff
@@ -0,0 +1,286 @@
+			   uintptr_t reg)
+{
+	uint32_t pin_mux, pin_cfg;
+	int ret = 0;
+
+	ARG_UNUSED(reg);
+
+	for (uint8_t i = 0U; i < pin_cnt; i++) {
+		pin_mux = pins[i].pinmux;
```
```suggestion
	ARG_UNUSED(reg);

	for (uint8_t i = 0U; i < pin_cnt; i++) {
		int ret;

		pin_mux = pins[i].pinmux;
```

glaubermaroto replied:
Though, in general, it is good practice to keep variables restricted to the scope they are being used. Here, I'd rather keep it out to avoid variable redefinitions or stack unwinding in different compiling optimization scenarios (including, eventually, debugging).

gmarull reviewed (dismissed):
lgtm, compliance check can be ignored (typedef)

glaubermaroto commented:
> @glaubermaroto you seem to have missed the comment that "the same check should be performed on all the driver changes in this PR"

@mbolivar-nordic It didn't go unnoticed, I did review the other drivers as well. Please, check my dismiss review message above.

glaubermaroto commented:
As stated before, the compliance check can be ignored.

@cfriedt Can you merge this one?

cfriedt commented:
> As stated before, the compliance check can be ignored.
> 
> @cfriedt Can you merge this one?

@glaubermaroto - I cannot. I only have merge privileges when all checks are passing. @nashif ? @carlescufi ? @stephanosio ?

glaubermaroto commented:
The last push was a rebase to solve conflicts.

glaubermaroto commented:
> > As stated before, the compliance check can be ignored.
> > @cfriedt Can you merge this one?
> 
> @glaubermaroto - I cannot. I only have merge privileges when all checks are passing. @nashif ? @carlescufi ? @stephanosio ?

@nashif ? @carlescufi ? @stephanosio ?
